File: examples/legacy/cs/gmconfig/gmconfig.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */






/*! \file gmconfig.cs
*
*  A C# example program demonstrating the use of a ConfigFile.
*  (Loading form a XML file, Creating a Connection, Subscribing & Publishing)
*/

using System;

namespace gmconfig
{
    using Nasa.Gmsec.Net;
    using Gmsec.Example.Common;

    class gmconfig : GmsecExample
    {
        private ConfigFile cfgFile;
        private Connection conn;
        private Message message;

        public gmconfig()
        {
        }

        private bool printUsage(int numberArgs)
        {
            if (numberArgs < 1)
            {
                Console.Out.WriteLine("\nusage: gmconfig_cs <filename>\n");
                return true;
            }
            return false;
        }

        private void Run(string[] args)
        {
            if (printUsage(args.Length))
            {
                return;
            }
            Status result = new Status();
            
            //Create the ConfigFile
            cfgFile = new ConfigFile(args[0]);

            //Load Config
            result = cfgFile.Load();
            check("ConfigFileLoad", result);

            Config config1;
            Config config2;
            result = cfgFile.LookupConfig("config1", out config1);
            check("LookupConfig", result);
            result = cfgFile.LookupConfig("config2", out config2);
            check("LookupConfig", result);
            
            //Dump Configs
            string xml;
            result = config1.ToXML(out xml);
            check("ToXml", result);
            Console.Out.WriteLine("Config 1=>\n " +xml);

            result = config2.ToXML(out xml);
            check("ToXml", result);
            Console.Out.WriteLine("Config 2=>\n " + xml);

            //Create Connection & Connect
            result = ConnectionFactory.Create(config2, out conn);
            check("Create Connection", result);
            result = conn.Connect();
            check("Connect", result);

            //LookUp Subscription & Subscribe
            string pattern1;
            string pattern2;
            result = cfgFile.LookupSubscription("events", out pattern1);
            check("LookUpSubscription", result);
            result = cfgFile.LookupSubscription("custom1", out pattern2);
            check("LookUpSubscription", result);
            Console.Out.WriteLine("Subscribing to: " + pattern1 + "\n");
            result = conn.Subscribe(pattern1);
            check("Subscribe", result);
            Console.Out.WriteLine("Subscribing to: " + pattern2 + "\n");
            result = conn.Subscribe(pattern2);
            check("Subscribe", result);
                
            //Create Message from Config
            result = conn.CreateMessage(out message);
            check("CreateMessage", result);
            result = cfgFile.LookupMessage("msg1", message);
            check("LookUpMessage", result);

            //Dump Message
            result = message.ToXML(out xml);
            check("ToXML", result);
            Console.Out.WriteLine("Message 1=>\n " + xml);
            //Publishing
            Console.Out.WriteLine("Publishing Message...");
            result = conn.Publish(message);
            check("Publish", result);
            
            //Clean UP
            result = conn.DestroyMessage(message);
            check("Destroy Message", result);
            Console.Out.WriteLine("Disconnecting...");
            result = conn.Disconnect();
            check("Disconnecting", result);
            result = ConnectionFactory.Destroy(conn);
            check("Destroying Connection", result);

        }

        static void Main(string[] args)
        {
            try
            {
                gmconfig exProg_gmconfig = new gmconfig();
                exProg_gmconfig.Run(args);
            }
            catch (Exception e)
            {
                Console.Out.WriteLine(e.ToString());
            }
        }
    }
}

File: examples/legacy/cpp/gmconfig/gmconfig.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */






/** 
 * @file gmconfig.cpp
 * 
 * A C++ example demonstration of loading a configuration from a standard
 * xml file.  The configuration file used for the example code is "gmconfig.xml".
 * 
 */

#include <gmsec_cpp.h>
#include <gmsec/util/sysutil.h>
#include "../example.h"
#include <iostream>
#include <string>

#include <gmsec/ConfigFileIterator.h>


using namespace std;
using namespace gmsec;
using namespace gmsec::util;
using namespace example;


class gmconfig {

public:

	gmconfig(ConfigFile &c);
	~gmconfig();

	Status Run();

private:

	ConfigFile &cf;
	Connection * connection;

};

gmconfig::gmconfig(ConfigFile &c) :
		cf(c), connection(0) {

	Config tmp;
	initialize(tmp);

}

gmconfig::~gmconfig() {

	if (connection) {

		if (connection->IsConnected())
			checkNoThrow("Disconnect", connection->Disconnect());

		checkNoThrow("Destroying Connection", ConnectionFactory::Destroy(connection));

	}

	checkNoThrow("ShutdownAllMiddlewares",
			ConnectionFactory::ShutdownAllMiddlewares());

}

Status gmconfig::Run() {

	Status result;

	/* Load and Parse */
	result = cf.Load();
	check("ConfigLoad", result);

	/* Retrieve 2 configs from file */
	Config c1, c2;
	cf.LookupConfig("config1", &c1);
	cf.LookupConfig("config2", &c2);

	/* Dump Details */
	const char *xml;
	result = c1.ToXML(xml);
	check("ToXML", result);

	LOG_INFO << "Config 1:" << xml;

	result = c2.ToXML(xml);
	check("ToXML", result);

	LOG_INFO << "Config 2:" << xml;

	/* Create the Connection */
	result = ConnectionFactory::Create(&c2, connection);
	check("Create Connection", result);

	/* Connect */
	result = connection->Connect();
	check("Connect", result);

	/* Suscribe by name */
	const char * pattern = NULL;
	result = cf.LookupSubscription("events", pattern);
	check("LookupSuscription", result);

	LOG_INFO << "Subscribing to: " << pattern;

	/* Create ConfigFileIterator & Reset */
	ConfigFileIterator cfgIter;
	cf.Reset(cfgIter);

	LOG_INFO << "Object names: ";

	LOG_INFO << "Subscription names: ";
	while(cfgIter.HasNextSubscriptionName())
	{
		gmsec::util::String gmstr;
		cfgIter.NextSubscriptionName(gmstr);
		std::string str(gmstr.c_str());
		LOG_INFO << str;
	}

	LOG_INFO << "Config names: ";
	while(cfgIter.HasNextConfigName())
	{
		gmsec::util::String gmstr;
		cfgIter.NextConfigName(gmstr);
		std::string str(gmstr.c_str());
		LOG_INFO << str;
	}
	
	LOG_INFO << "Message names: ";
	while(cfgIter.HasNextMessageName())
	{
		gmsec::util::String gmstr;
		cfgIter.NextMessageName(gmstr);
		std::string str(gmstr.c_str());
		LOG_INFO << str;
	}

	LOG_INFO << "Subscriptions from list: ";
	while(cfgIter.HasNextSubscription())
	{
		gmsec::util::String gmstr;
		cfgIter.NextSubscription(gmstr);
		std::string str(gmstr.c_str());
		LOG_INFO << str;
	}

	LOG_INFO << "Configs from list: ";
	while(cfgIter.HasNextConfig())
	{
		const char* xml;
		Config* cfg = new Config();
		cfgIter.NextConfig(cfg);
		cfg->ToXML(xml);
		LOG_INFO << xml;
		delete cfg;
	}

	LOG_INFO << "Messages from list: ";
	while(cfgIter.HasNextMessage())
	{
		const char* xml;
		Message* msg;
		connection->CreateMessage("TEMP",GMSEC_MSG_PUBLISH, msg);
		cfgIter.NextMessage(msg);
		msg->ToXML(xml);
		LOG_INFO << xml;
		connection->DestroyMessage(msg);
	}
	

	result = connection->Subscribe(pattern);
	check("Subscribing", result);

	/* Create message from config entry */
	Message * message = 0;
	result = connection->CreateMessage(message);
	check("CreateMessage", result);

	result = cf.LookupMessage("msg1", message);
	check("LookupMessage", result);

	/* Dump Message */
	result = message->ToXML(xml);
	check("ToXML", result);
	LOG_INFO << "Message 1: " << xml;

	/* Publish Message */
	result = connection->Publish(message);
	check("Publish", result);

	/* Destroy the Message */
	result = connection->DestroyMessage(message);
	check("DestroyMessage", result);

	return result;

}

int main(int argc, char* argv[]) {

	if (argc <= 1) {

		cout << "usage: gmconfig <filename>" << endl;
		return -1;

	}

	ConfigFile cf(argv[1]);

	try {

		gmconfig state(cf);
		state.Run();

	} catch (exception &e) {

		LOG_ERROR << "Failed: " << e.what();
		return -1;

	}

	return 0;

}
